### ISSUE TITLE
IfcPatch ConvertLengthUnit recipe doesn't convert IfcCurveSegment.SegmentStart and SegmentLength

### ISSUE THREAD

**Issue by RickBrice** [IfcOpenShell/IfcOpenShell#5656]
I am trying to use IfcPatch ConvertLengthUnit recipe to convert the files in https://github.com/bSI-RailwayRoom/IFC-Rail-Unit-Test-Reference-Code/tree/master/alignment_testset/IFC-WithGeneratedGeometry to FOOT units.

Everything works great except for the SegmentStart and SegmentLength attributes of IfcCurveSegment.

This seems like a good "beginner" problem to solve, so I'm looking for some guidance on how to approach the problem and develop a solution.

It looks like the work is done in `ifcopenshell.util.unit.convert_file_length_units` which calls 

https://github.com/IfcOpenShell/IfcOpenShell/blob/3d4988349af0c0ce11591d2482b58c999398728e/src/ifcopenshell-python/ifcopenshell/util/unit.py#L827-L834

For one of the files I'm trying to convert the SegmentLength attribute is IFCLENGTHMEASURE(100.0):

> #36 = IFCCURVESEGMENT(.CONTINUOUS., #42, IFCLENGTHMEASURE(0.), IFCLENGTHMEASURE(100.), #45);

It seems like the IFCLENGTHMEASURE should be converted, but it doesn't. Perhaps it has to do with mixed case versus all upper case text.

I would like guidance on how to solve this myself. I am willing to get onto OSArch for a screen share if needed.

**Comment by aothms** [comment]
I began looking into this and couldn't stop.

I'm amazed we even have this functionality. I built the same for the rule execution, but using more like a callback rather than an iterator. https://github.com/IfcOpenShell/IfcOpenShell/blob/3d4988349af0c0ce11591d2482b58c999398728e/src/ifcopenshell-python/ifcopenshell/express/rule_executor.py#L231-L238

Below fixes things for me, but still needs to be clean up, gotta run. This requires quite in depth knowledge of our internal representation of the schema which is fully undocumented and a bit awkward. Especially the distinction between named_type and type_declaration. But by making it a tad bit more recursive the function is simplified a bit.

```python
def is_attr_type(
    content_type: ifcopenshell_wrapper.parameter_type, ifc_unit_type_name: str, include_select_types: bool = True
) -> Union[ifcopenshell_wrapper.type_declaration, None]:
    cur_decl = content_type

    if hasattr(cur_decl, "name") and cur_decl.name() == ifc_unit_type_name:
        return cur_decl

    if include_select_types:
        if hasattr(cur_decl, "select_list"):
            for select_item in cur_decl.select_list():
                if is_attr_type(select_item, ifc_unit_type_name):
                    return select_item

    # while hasattr(cur_decl, "declared_type") is True:
    #     cur_decl = cur_decl.declared_type()
    #     if include_select_types:
    #         if hasattr(cur_decl, "select_list"):
    #             for select_item in cur_decl.select_list():
    #                 if is_attr_type(select_item, ifc_unit_type_name):
    #                     return select_item
    #     if hasattr(cur_decl, "name") is False:
    #         continue
    #     if cur_decl.name() == ifc_unit_type_name:
    #         return cur_decl
                
    if hasattr(cur_decl, "declared_type"):
        return is_attr_type(cur_decl.declared_type(), ifc_unit_type_name, include_select_types)

    if isinstance(cur_decl, ifcopenshell_wrapper.aggregation_type):
        # support aggregate of aggregates, as in IfcCartesianPointList3D.CoordList
        def get_declared_type_from_aggregate(cur_decl):
            cur_decl = cur_decl.type_of_element()
            if not isinstance(cur_decl, ifcopenshell_wrapper.aggregation_type):
                return cur_decl.declared_type()
            return get_declared_type_from_aggregate(cur_decl)

        cur_decl = get_declared_type_from_aggregate(cur_decl)
        return is_attr_type(cur_decl, ifc_unit_type_name, include_select_types)
        """
        if hasattr(cur_decl, "name") and cur_decl.name() == ifc_unit_type_name:
            return cur_decl
        while hasattr(cur_decl, "declared_type") is True:
            cur_decl = cur_decl.declared_type()
            if hasattr(cur_decl, "name") is False:
                continue
            if cur_decl.name() == ifc_unit_type_name:
                return cur_decl
        """
    return None
```

**Comment by aothms** [comment]
![1-s2 0-S0926580519311288-gr4](https://github.com/user-attachments/assets/93438775-cf9b-4199-86ee-120d7cff9229)

I just remembered I have this. I think I created this, at least its in one of my papers. https://www.sciencedirect.com/science/article/pii/S0926580519311288#f0020

Not sure if this diagram is completely followed in the schema representation embedded in ifopsh

**Comment by RickBrice** [comment]
> This seems like a good "beginner" problem to solve

Or not?

Thanks @aothms but this is way over my head - I'm not sure what you are trying to tell me.

If you've got a fix, please implement it. Thanks.

**Comment by RickBrice** [comment]
Fixed by @aothms via 4cd8428d1de496bf2e23570055bcde81bad1e03e
